java/arcs/core/data/proto/AnnotationProtoDecoder.kt

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.proto
+
+import arcs.core.data.Annotation
+import arcs.core.data.AnnotationParam
+
+fun AnnotationParamProto.decode(): AnnotationParam {
+    return when (valueCase) {
+        AnnotationParamProto.ValueCase.STR_VALUE -> AnnotationParam.Str(strValue)
+        AnnotationParamProto.ValueCase.NUM_VALUE -> AnnotationParam.Num(numValue)
+        AnnotationParamProto.ValueCase.BOOL_VALUE -> AnnotationParam.Bool(boolValue)
+        else -> throw IllegalArgumentException("Cannot decode a [AnnotationParamProto].")
```

Comment:
Might be nice to specify in this error message that the type was not known. Here's an example: https://github.com/PolymerLabs/arcs/blob/master/java/arcs/android/crdt/CrdtEntityProto.kt#L65-L67

Additionally, and this is a minor nit, but an UnsupportedOperationException is probably more appropriate here. It's a subtle distinction, but IllegalArgumentException means "you gave me a bad argument". UnsupportedOperationException means "what you gave me / tried to do was valid, but I don't support it". If we add a new type of annotation param, but don't update this code, then it will appear as the latter case: you tried to use a real annotation param type, but this code doesn't handle it correctly.

Reply:
sorry, i keep using the wrong exceptions, thanks for putting this into words for me, i think it is clearer now, i'll try harder next time!

Reply:
No worries! It took me a while to learn too. My rules of thumb:

* IllegalArgumentException: what you gave me was invalid.
* IllegalStateException: my state is messed up, or the state I'm in means I can't do what you asked.
* NoSuchElementException: the thing you asked for doesn't exist.
* UnsupportedOperationException: I'm sorry, Dave. I'm afraid I can't do that.